File: hackerrank/problem solving/algorithms/2 - implementation/27 - sequence equation/sequence_equation.cs
// https://www.hackerrank.com/challenges/permutation-equation/problem?isFullScreen=true

Console.ReadLine();
int[] arr = ReadNumbers();
foreach (var x in SequenceEquation(arr))
    Console.WriteLine(x);

int[] ReadNumbers() => Console.ReadLine()!.Split().Select(int.Parse).ToArray();

// n: length of arr
// T: O(n)
// S: O(n) extra space
IEnumerable<int> SequenceEquation(int[] arr)
{
    int[] valuesToIndex = new int[arr.Length];
    int i = 0;
    foreach (int val in arr)
        valuesToIndex[val - 1] = i++;
    foreach (int val in valuesToIndex)
        yield return valuesToIndex[val] + 1;
}

File: hackerrank/problem solving/algorithms/2 - implementation/27 - sequence equation/sequence_equation.java
// https://www.hackerrank.com/challenges/permutation-equation/problem?isFullScreen=true
// Java 25

void main() {
    IO.readln();
    int[] arr = readNumbers();
    for (int x : sequenceEquation(arr))
        IO.println(x);
}

int[] readNumbers() {
    return Stream.of(IO.readln().split(" ")).mapToInt(Integer::parseInt).toArray();
}

// n: length of array arr/output
// T: O(n)
// S: O(n) extra space
int[] sequenceEquation(int[] arr) {
    int[] valuesToIndex = new int[arr.length];
    int i = 0;
    for (int val : arr) {
        valuesToIndex[val - 1] = i++;
    }
    return Arrays.stream(valuesToIndex).map(val -> valuesToIndex[val] + 1).toArray();
}
